raysect/core/math/function/vectorfunction2d/base.pyx

```diff
@@ -0,0 +1,235 @@
+
+import numbers
+from cpython.object cimport Py_LT, Py_EQ, Py_GT, Py_LE, Py_NE, Py_GE
+cimport cython
+from libc.math cimport floor
+from raysect.core.math.vector cimport Vector3D
+from raysect.core.math.function.function2d.base cimport Function2D
+from raysect.core.math.function.function2d.autowrap cimport autowrap_function2d
+from raysect.core.math.function.function2d.base cimport is_callable as f2d_is_callable
+from .autowrap cimport autowrap_vectorfunction2d
+
+
+cdef class VectorFunction2D:
+    """
+    Cython optimised class for representing an arbitrary 2D vector function.
+
+    Using __call__() in cython is slow. This class provides an overloadable
+    cython cdef evaluate() method which has much less overhead than a python
+    function call.
+
+    For use in cython code only, this class cannot be extended via python.
+
+    To create a new function object, inherit this class and implement the
+    evaluate() method. The new function object can then be used with any code
+    that accepts a function object returning a Vector3D.
+    """
+
+    cdef Vector3D evaluate(self, double x, double y):
+        raise NotImplementedError("The evaluate() method has not been implemented.")
+
+    def __call__(self, double x, double y):
+        """ Evaluate the function f(x, y)
+
+        :param float x: function parameter x
+        :param float y: function parameter y
+        :rtype: float
+        """
+        return self.evaluate(x, y)
+
+    def __add__(object a, object b):
+        if is_callable(a) or isinstance(a, Vector3D):
+            if is_callable(b) or isinstance(b, Vector3D):
+                return AddVectorFunction2D(a, b)
+        return NotImplemented
+
+    def __sub__(object a, object b):
+        if is_callable(a) or isinstance(a, Vector3D):
+            if is_callable(b) or isinstance(b, Vector3D):
+                return SubtractVectorFunction2D(a, b)
+        return NotImplemented
+
+    def __mul__(object a, object b):
+        if is_callable(a) or isinstance(a, Vector3D):
+            if f2d_is_callable(b) or isinstance(b, numbers.Real):
+                return MultiplyVectorFunction2D(a, b)
+        if is_callable(b) or isinstance(b, Vector3D):
+            if f2d_is_callable(a) or isinstance(a, numbers.Real):
+                return MultiplyVectorFunction2D(b, a)
+        return NotImplemented
+
+    def __truediv__(object a, object b):
+        if is_callable(a) or isinstance(a, Vector3D):
+            if f2d_is_callable(b) or isinstance(b, numbers.Real):
+                return DivideVectorFunction2D(a, b)
+        return NotImplemented
+
+    def __neg__(self):
+        return NegVectorFunction2D(self)
+
+    def __richcmp__(self, object other, int op):
+        if is_callable(other) or isinstance(other, Vector3D):
+            if op == Py_EQ:
+                return EqualsVectorFunction2D(self, other)
+            if op == Py_NE:
+                return 1 - EqualsVectorFunction2D(self, other)
```

Comment:
I'd prefer an explicit NotEqual class for this. Otherwise the user gets back an uninformative SubtractVectorFunction2D object.

It does occur to me that we are abusing the comparison operators here, They are meant to return True/False, not an object than returns True/False when called. This excursion from expected behaviour needs documenting very clearly. I think I missed this when looking through the original function3D implementation.

Reply:
If I remember rightly, we discussed this in the `Function2D` implementation and came to the conclusion that it was fine, because a) `double(True) == 1.0` and `double(False) == 0.0` is explicitly defined by CPython (or Cython, I forget which) and b) the tests compare the results to 1.0 and 0.0, so any change would result in a failing test anyway. That said, I'm happy to do the cast explicitly:
```
return 1.0 if function1.evaluate(x, y) == function2.evaluate(x, y) else 0.0
```
I don't think this will have a performance impact: the compiler will probably end up emitting the same instructions as for the implicit cast.

Reply:
I'm not sure if you saw I edited my comment (not sure it emails you if I do). I ended up editing my thoughts as I was too quick to comment.

I would prefer an explicit object as stated above. It would also be helpful if I implement a __repr__ or `inspect()` system to display the chain of equations.

Reply:
I've now added a separate `NotEqualsVectorFunction2D` class, and modified the evaluate method of both classes to explicitly return either 1.0 or 0.0.

The Python docs state that "by convention" the `__eq__` and `__neq__` methods return true or false, but doesn't state that they should. I think of this function framework as an expression building engine, and the expression is only evaluated at call time. So I don't think that the comparison methods returning non-boolean objects is a problem: the objects they return aren't meant to be treated as values anyway. This is consistent with the rest of the framework. For example, `f1 + f2 - f2` does not give an object identical to `f1`: the fact that subtraction is the inverse of addition is not satisfied until the object is evaluated.